community/kernel/src/main/java/org/neo4j/kernel/impl/logging/StoreLogService.java

```diff
@@ -149,56 +167,81 @@ private StoreLogService( LogProvider userLogProvider,
             internalLogProvider = internalLogBuilder.toOutputStream( outputStream );
-            rotationListener.accept( internalLogProvider );
-            this.closeable = outputStream;
+            builder.rotationListener.accept( internalLogProvider );
+
+            // We manage closing of the internal log since we create it in here. The user log is provided
+            // to us (no pun intended) so we'll let the provider take care of closing that.
+            closeInShutdown = outputStream;
         }
         else
         {
-            RotatingFileOutputStreamSupplier rotatingSupplier = new RotatingFileOutputStreamSupplier( fileSystem, internalLog,
-                    internalLogRotationThreshold, internalLogRotationDelay, maxInternalLogArchives,
-                    rotationExecutor, new RotatingFileOutputStreamSupplier.RotationListener()
+            RotatingFileOutputStreamSupplier rotatingSupplier = new RotatingFileOutputStreamSupplier(
+                    fileSystem,
+                    internalLog,
+                    builder.internalLogRotationThreshold,
+                    builder.internalLogRotationDelay,
+                    builder.maxInternalLogArchives,
+                    builder.rotationExecutor,
+                    new RotatingFileOutputStreamSupplier.RotationListener()
             {
                 @Override
                 public void outputFileCreated( OutputStream newStream, OutputStream oldStream )
                 {
                     FormattedLogProvider logProvider = internalLogBuilder.toOutputStream( newStream );
                     logProvider.getLog( StoreLogService.class ).info( "Opened new internal log file" );
-                    rotationListener.accept( logProvider );
+                    builder.rotationListener.accept( logProvider );
                     logProvider.getLog( StoreLogService.class ).info( "Rotated internal log file" );
                 }
             } );
             internalLogProvider = internalLogBuilder.toOutputStream( rotatingSupplier );
-            this.closeable = rotatingSupplier;
+            closeInShutdown = rotatingSupplier;
+        }
+
+        // Configure the log providers to be asynchronous, if user wanted that behavior.
+        LogProvider userLogProvider = builder.userLogProvider;
+        if ( builder.asyncExecutor != null )
+        {
+            // Wraps each log provider in an AsyncLogProvider and schedules the flushing of log events in an executor,
+            // where the scheduling into the executor takes place during init().
+            // Also Reassign these log providers so that the async versions are used below.
+            userLogProvider = internalLife.add( new AsyncLogProvider( userLogProvider, builder.asyncExecutor ) );
+            internalLogProvider = internalLife.add( new AsyncLogProvider( internalLogProvider, builder.asyncExecutor ) );
```

Comment:
This looks like we will get two different log processing threads, one for each of the two logs we have.
Since we (most likely) only have one I/O subsystem, would it not be better to process both event streams on the same thread?

That would mean enqueueing the events for both logs into the same AsyncEvents queue (since the events know which log they are writing to already, this should be fine).